lib/HttpClient.ts

```diff
@@ -226,63 +247,93 @@ export class HttpClient {
+    /**
+     * Raw request.
+     * @param info 
+     * @param data 
+     */
+    public requestRaw(info: ifm.IRequestInfo, data: string | NodeJS.ReadableStream): Promise<ifm.IHttpClientResponse> {
+        return new Promise<ifm.IHttpClientResponse>((resolve, reject) => {
+            var callbackForResult = function (err: any, res: ifm.IHttpClientResponse) {
+                if (err) {
+                    reject(err);
+                }
 
-            let isDataString = typeof (data) === 'string';
+                resolve(res);
+            };
 
-            if (typeof (data) === 'string') {
-                info.options.headers["Content-Length"] = Buffer.byteLength(data, 'utf8');
-            }
+            this.requestRawWithCallback(info, data, callbackForResult);
+        });
+    }
 
-            let req: http.ClientRequest = info.httpModule.request(info.options, (msg: http.IncomingMessage) => {
-                let res: HttpClientResponse = new HttpClientResponse(msg);
-                resolve(res);
-            });
+    /**
+     * Raw request with callback.
+     * @param info 
+     * @param data 
+     * @param onResult 
+     */
+    public requestRawWithCallback(info: ifm.IRequestInfo, data: string | NodeJS.ReadableStream, onResult: (err: any, res: ifm.IHttpClientResponse) => void): void {
+        let socket;
+        
+        let isDataString = typeof (data) === 'string';
+        if (typeof (data) === 'string') {
+            info.options.headers["Content-Length"] = Buffer.byteLength(data, 'utf8');
+        }
 
-            req.on('socket', (sock) => {
-                socket = sock;
-            });
+        var callbackCalled: boolean = false;
+        var handleResult = (err: any, res: HttpClientResponse) => {
+            if (!callbackCalled) {
+                callbackCalled = true;
+                onResult(err, res);
+            }
+        };
 
-            // If we ever get disconnected, we want the socket to timeout eventually
-            req.setTimeout(this._socketTimeout || 3 * 60000, () => {
-                if (socket) {
-                    socket.end();
-                }
-                reject(new Error('Request timeout: ' + info.options.path));
-            });
+        let req: http.ClientRequest = info.httpModule.request(info.options, (msg: http.IncomingMessage) => {
+            let res: HttpClientResponse = new HttpClientResponse(msg);
+            handleResult(null, res);
+        });
 
-            req.on('error', function (err) {
-                // err has statusCode property
-                // res should have headers
-                reject(err);
-            });
+        req.on('socket', (sock) => {
+            socket = sock;
+        });
 
-            if (data && typeof (data) === 'string') {
-                req.write(data, 'utf8');
+        // If we ever get disconnected, we want the socket to timeout eventually
+        req.setTimeout(this._socketTimeout || 3 * 60000, () => {
```

Comment:
is this 3 minutes? can you add a comment

Reply:
i now see this is largely existing code with the addition of handleResult

what problem existed that you are solving with handleResult? reject called twice? would be great to have some pr comments for the reader explaining this. it takes a while to attempt to infer from the diff, especially since this method is largely syntax changes and the diff is hard to read.

Reply:
what was the symptom before? did we have a case where it manifested in strange behavior? or did you just notice a bug?

Reply:
Will add a comment for the default timeout.

Reply:
The formatting of the diff got really messed up, it looks terrible.

This change is due to adding NTLM. The way keep alive connections work is that a subsequent request has to be queued on the socket when the first request finishes otherwise the socket closes. When using async/await I kept getting an HTTP 401 Unauthorized. My thought was that the way I was using setImmediate in conjunction with async/await caused the second request to happen too late in the node event loop so the socket was closing. For NTLM if the request is on a different connection then it fails, both NTLM messages have to be on the same connection.

In order to fix this I refactored part of the code to use callbacks so that I could call setImmediate inside the callback of the first request. I believe this changes when in the node event loop the request gets queued and ensures that the second request is pushed to the connection request queue before the nextTick in the event loop(this must be when the connection checks if anything is queued and drops the connection if the request queue is empty). This allows the requests to both happen on the same connection and the NTLM handshake to be successful.

I will note that it's possible I wasn't doing setImmediate in conjunction with async/await correctly but I messed around with it for a while and couldn't get it working and the other code we copied from to do NTLM all uses callbacks. I was seeing some strangeness with readBody on the response message that also caused a 401 so it's possible I was conflating the two issues and if ere readBody is done correctly and the code uses async/await  it works but I am not sure. If it makes sense I can try to follow this path and see. Unfortunately testing NTLM is very time consuming and tricky to debug.